File: 54-spiral-matrix/spiral-matrix.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if(matrix.size() == 0)
            return {};
        int m=matrix.size();
        int n=matrix[0].size();
        int dir=0;
        int top   = 0;
        int down  = m-1;
        int left  = 0;
        int right = n-1;
        vector<int> res;
        while(top <= down && left <= right){
            if(dir==0){
                //left to right
                for(int i=left;i<=right;i++){
                    res.push_back(matrix[top][i]);
                }
                top++;
            }
            if(dir==1){
                //top to down
                for(int i=top;i<=down;i++){
                    res.push_back(matrix[i][right]);
                }
                right--;
            }
            if(dir==2){
                //right to left
                for(int i=right;i>=left;i--){
                    res.push_back(matrix[down][i]);
                }
                down--;
            }
            if(dir==3){
                //down to top
                for(int i=down;i>=top;i--){
                    res.push_back(matrix[i][left]);
                }
                left++;
            }
            dir = (dir+1)%4;
        }
        return res;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []
        res=[]
        left,right=0,len(matrix[0])
        top,bottom=0,len(matrix)
        while(left<right and top<bottom):
            for j in range(left,right):
                res.append(matrix[top][j])
            top+=1
            for i in range(top,bottom):
                res.append(matrix[i][right-1])
            right-=1
            if not(left<right and top<bottom):
                break
            for j in range(right-1,left-1,-1):
                res.append(matrix[bottom-1][j])
            bottom-=1
            for i in range(bottom-1,top-1,-1):
                res.append(matrix[i][left])
            left+=1
        return res
            
# class Solution:
#     def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
#         if not matrix:
#             return []
#         rowbegin=0
#         rowend=len(matrix)
#         colbegin=0
#         colend=len(matrix[0])
#         res=[]
#         while rowend>rowbegin and colend>colbegin:
#             for i in range(colbegin,colend):
#                 res.append(matrix[rowbegin][i])
#             for i in range(rowbegin+1,rowend-1):
#                 res.append(matrix[i][colend-1])
#             if rowend !=rowbegin+1:
#                 for i in range(colend-1,colbegin-1,-1):
#                     res.append(matrix[rowend-1][i])
#             if colbegin!=colend-1:
#                 for i in range(rowend-2,rowbegin,-1):
#                     res.append(matrix[i][colbegin])
#             rowbegin+=1
#             rowend-=1
#             colbegin+=1
#             colend-=1
#         return res